+frames/Index.m

```
@@ -234,47 +302,345 @@
+            
+            
+            function out = labelDuplicatesInSections(x, L, unique_section)
+                % get label vector for duplicate values of x, per section defined by lengths in L
+                % (for speedup: do not calculate if section is known to be unique)
+                startpos=[1 cumsum(L)+1];
+                out = zeros(size(x));
+                for i=1:length(L)
+                    p1 = startpos(i);
+                    p2 = startpos(i+1)-1;
+                    if unique_section(i)
+                        out(p1:p2) = 1;
+                    else
+                        out(p1:p2) = labelDuplicates(x(p1:p2));
+                    end
+                end                
+            end            
+            
+            function out = labelDuplicates(x)
+                % label duplicate values in array
+                [X, ind_sort] = sort(x);
+                C = countWithReset([1; diff(X)]);
+                [~, ind_revsort] = sort(ind_sort);
+                out = C(ind_revsort);
+            end
+            
+            function out = countWithReset(v)
+                % count consecutive zero elements, reset count at other values
+                count = 1;
+                out = zeros(size(v));
+                for i=1:length(v)
+                    if v(i)
+                        count = 1;
+                    else
+                        count = count + 1;
+                    end
+                    out(i) = count;
+                end
+            end
+        end
+        
+        
+        function [objnew, ind1_new, ind2_new] = alignIndex(obj1, obj2, method, ~)
+            % function to create new aligned Index of two Index objects            
+            %
+            % input:
+            %    - obj1,obj2:    Index objects to be aligned
+            %
+            %    - method: (string enum) select method
+            %           "strict": both need to have same values (else error thrown)
```

I think that the method should return an error with duplicates, unless they are exactly equal (in the same order). Otherwise you can get unexpected behavior with elementWise operations and aligment.

Pandas duplicates the index for elementWise operations (if you have idx1 = [3,3], idx2=[3,3] -> idxOut = [3,3,3,3]. And fails when trying to concatenate DataFrames with duplicated indices (`InvalidIndexError: Reindexing only valid with uniquely valued Index objects`).

---

Good remark, it seems indeed that the alignment behavior with duplicates is not implemented correctly. 

If the indices are exactly the same, they are just 1:1 mapped.  This is same behavior as the current frames. But when the indices are not completely equal, the result does not align the duplicates currently. The are just ignored (made NaN).

I agree that a solution that assumes the duplicates are ordered the same way (when index is not exactly the same)  is dangerous. Expanding (like pandas) to all the combinations is I think  not desirable for most use-cases, and not consistent with the current behavior (when indices are equal). Throwing an error is indeed most logical solution.

Remark: 
The auto-expansion (in case of elementwise operations with MultiIndex with different dimension names) does expands to all the combinations including duplicates.

---

To get the required behavior I splitted the behavior for duplicates from the alignMethod setting. 

duplicateOptions determines behavior with duplicates in the alignment between 2 dataframes, alignMethod determines what unique values are kept in the indices from both dataframes.


         alignMethod (1,1) string {mustBeMember(alignMethod,["strict", "inner", "left", "full","strictunique"])} = "strict"
                             % defines how indices are aligned in math operation between DataFrames:
                             % -  "strict":  both need to have same indices (else error thrown)
                             % -  "inner":   remove index values that are not common in both
                             % -  "left":    keep indices as in left DataFrame in operation (default)            
                             % -  "full":    keep all indices values (allow missing values in indices in both)
                       
        duplicateOption (1,1) string ...
                        {mustBeMember(duplicateOption,["none", "unique", "duplicates","duplicatesstrict","expand"])} ...
                        = "duplicatesstrict"
                             % defines how duplicate values in indices are handled in math operation on DataFrames:                             
                             %  - "none":             no operations with duplicates allowed
                             %  - "unique":           removes duplicates, only use first occurrence of duplicate value
                             %  - "duplicatestrict":  only allow duplicates in case of equal indices
                             %  - "duplicates":       align duplicates in order of occurrence between indices
                             %                        (first occurrence duplicate value in one index is matched to 
                             %                         1st occurence in other, 2nd occurence to 2nd etc)
                             %  - "expand":           create all combinations of matching duplicates between indices
                             %                        in `outputs


The 'duplicatestrict' default duplicationOption now make sure that duplicates are only allowed if the indices are exactly the same. An expand option is now also available to get the behavior as in pandas.